crates/settings/src/blocks.rs

```diff
@@ -0,0 +1,372 @@
+
+impl<'de> Visitor<'de> for BlockRangeVisitor {
+    type Value = BlockRange;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a range in the form [a..b], [a..], or [..b]")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        parse_range(v).map_err(de::Error::custom)
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: de::SeqAccess<'de>,
+    {
+        let mut range_string = String::new();
+
+        while let Some(elem) = seq.next_element::<String>()? {
+            range_string.push_str(&elem);
+        }
+
+        parse_range(&range_string).map_err(de::Error::custom)
+    }
+}
+
+fn parse_range(s: &str) -> Result<BlockRange, String> {
+    let parts: Vec<&str> = s.split("..").collect();
+    if parts.len() == 2 {
+        let start = match parts[0] {
+            "" => Block::Genesis,
+            s => Block::Number(s.parse().map_err(|_| "Invalid block number")?),
+        };
+        let end = match parts[1] {
+            "" => Block::Latest,
+            s => Block::Number(s.parse().map_err(|_| "Invalid block number")?),
+        };
+        return Ok(BlockRange { start, end });
+    }
+    Err(format!("Invalid range syntax: {}", s))
+}
+
+#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
+#[serde(untagged)]
+#[typeshare]
+pub enum Blocks {
+    RangeWithInterval { range: BlockRange, interval: u32 },
+    SimpleRange(BlockRange),
+}
+
+#[derive(Debug, Error, PartialEq)]
+pub enum BlockError {
+    #[error("Invalid block range")]
+    InvalidBlockRange,
+}
+
+impl Blocks {
+    pub fn expand_to_block_numbers(
+        &self,
+        latest_block: BlockNumber,
+    ) -> Result<Vec<BlockNumber>, BlockError> {
+        match self {
+            Blocks::RangeWithInterval { range, interval } => {
+                range.validate(latest_block)?;
+                let mut blocks = vec![];
+                let mut current_block = range.start.to_block_number(latest_block);
+                let end_block = range.end.to_block_number(latest_block);
+                while current_block <= end_block {
+                    blocks.push(current_block);
+                    current_block += *interval as BlockNumber;
+                }
+                Ok(blocks)
+            }
+            Blocks::SimpleRange(range) => {
+                range.validate(latest_block)?;
+                let start_block = range.start.to_block_number(latest_block);
+                let end_block = range.end.to_block_number(latest_block);
+                Ok((start_block..=end_block).collect())
+            }
+        }
+    }
+}
+
+impl Default for Blocks {
+    fn default() -> Self {
+        Blocks::RangeWithInterval {
+            range: BlockRange {
+                start: Block::Genesis,
+                end: Block::Latest,
+            },
+            interval: 1,
```

Comment:
hmm, a safer default would be start/end both at latest, this will expand out to millions of iterations probably

Reply:
ah yeah fair, i put this in for other tests where we need a default, but actually it's an `Option` in Scenario anyway, so I can delete this.